janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/AdjacentVertexHasUniquePropertyOptimizerStrategy.java

```diff
@@ -0,0 +1,145 @@
+        if (TraversalHelper.onGraphComputer(traversal)
+            || traversal.getStartStep() instanceof TraversalVertexProgramStep // not caught by TraversalHelper
+            || !traversal.getGraph().isPresent()) {
+            return;
+        }
+
+        Graph graph = traversal.getGraph().get();
+        StandardJanusGraph janusGraph = graph instanceof StandardJanusGraph
+            ? (StandardJanusGraph) graph
+            : ((StandardJanusGraphTx) graph).getGraph();
+
+        if (!janusGraph.getConfiguration().optimizerBackendAccess()) {
+            return;
+        }
+
+        TraversalHelper.getStepsOfClass(HasStep.class, traversal)
+            .forEach(this::optimizeStep);
+    }
+
+    /**
+     * Determines whether this HasStep can be answered by a unique index and thus, returns either 0 or 1 match
+     */
+    @Override
+    protected boolean isValidStep(HasStep<?> step) {
+        StandardJanusGraphTx tx = (StandardJanusGraphTx) JanusGraphTraversalUtil.getTx(step.getTraversal());
+
+        List<String> givenKeys = step.getHasContainers().stream()
+            .map(HasContainer::getKey).collect(Collectors.toList());
+
+        List<PredicateCondition<String, JanusGraphElement>> constraints = step.getHasContainers().stream()
+            .filter(hc -> hc.getBiPredicate() == Compare.eq)
+            .map(hc -> new PredicateCondition<>(hc.getKey(), JanusGraphPredicateUtils.convert(hc.getBiPredicate()), hc.getValue()))
+            .filter(pc -> tx.validDataType(pc.getValue().getClass()))
+            .collect(Collectors.toList());
+        final MultiCondition<JanusGraphElement> conditions = QueryUtil.constraints2QNF(tx, constraints);
+
+        // collect all matching indexes
+        final Set<IndexType> availableIndexes = IndexSelectionUtil.getMatchingIndexes(conditions);
+
+        // select only valid unique indexes
+        return availableIndexes.stream()
+            .filter(IndexType::isCompositeIndex)
+            .map(index -> (CompositeIndexType) index)
+            .filter(cIndex -> cIndex.getCardinality() == Cardinality.SINGLE)
+            .anyMatch(cIndex -> IndexSelectionUtil.isIndexSatisfiedByGivenKeys(cIndex, givenKeys));
+    }
+
+    private Traversal.Admin<?,Long> generateFilter(Traversal.Admin<?,?> traversal, HasStep<?> originalStep) {
+        Traversal.Admin filterTraversal = new DefaultGraphTraversal<>(traversal.getGraph().get())
+            .addStep(
+                new GraphStep<>(traversal, Vertex.class, true)
+            );
+
+        filterTraversal.addStep(originalStep);
+        filterTraversal.addStep(new IdStep<>(filterTraversal));
+        return filterTraversal;
+    }
+
+    private long evaluateFilter(Traversal.Admin<?,Long> filterTraversal) {
+        // negative IDs are not valid in JanusGraph. Therefore, if the filterTraversal has no result, the newly
+        // inserted HasStep should not match anything
+        long determinedId = -1;
```

Comment:
This does not work due to https://github.com/JanusGraph/janusgraph/blob/4f44a1db63513f01cd118b6ccbe24b5cdb36878d/janusgraph-core/src/main/java/org/janusgraph/graphdb/query/vertex/BaseVertexCentricQueryBuilder.java#L107-L111

You can see this problem by adding a test case where the adjacent vertex does not exist, e.g. `assertNumStep(1, 0, g.V(sv[0]).out().has("uniqueId", 10000), HasStep.class);`

Reply:
Crucial! I think I found a better solution now, since I discovered Gremlin supports a `none()` step, which is exactly what I tried to emulate here with `has('~adjacent', -1)` but with style :grin: 

I also included your test case (which should be `assertNumStep(0, 0, ...`, right?) and will now wait for the build to have look at the code coverage and see where I need additional tests. Thanks for taking the time once again to review my PR!